### ISSUE TITLE
needs optimisation

### ISSUE THREAD

**Issue by alxndrwllmsn** [rstofi/gridflag#1]
This is mostly for my benefit, while I work on optimising the code. If you have any comments that you want to add as we go, put them here.

The summary from our email is as such:

- creating a jupyter notebook would be useful for testing
- disable regionRadiusSq computations to speed up testing but it's probably needed in the long-run

My starting point will probably be to profile the code and convert your depth-4 loops into numpy array calculations.
I'll add more comments here as I go.

**Comment by alxndrwllmsn** [comment]
@rstofi are there any cases where boxWidth is different for different channels? If we can make the assumption that it is identical for each channel, we should be able to vectorise the whole operation. Otherwise, we'll have to do some magic to make numpy think that the array is the same size for all channels.

**Comment by rstofi** [comment]
@alxndrwllmsn possibly good to discuss issues here.

I have 4 nested loops mainly because I wanted to follow the original code as much as possible.

The top loop is obviously through channels. We should obviously parallelise this, I just had no time to do that for testing the code...
For each channel the `boxWidth` could be different in theory, however, it is expected to be the same. I have some code commented out that gives you `boxWidth` over the frequency axis:

```
pcf_kernel_sizes = np.fabs(np.divide(pcf_imag,pcf_real, where=pcf_real!=0))
C_max_array = np.ceil(np.amax(pcf_kernel_sizes, axis=(1,2))) 
#Each element equals the boxWidth for the corresponding channel
```

The remaining loops are:

 - over the image pixels
 - over the area of `boxWidth` centred at (x,y) to get the max local kernel size (this could be easily vectorised via scipy)
 - over the area of the local largest kernel (`kernelW`) centred at (x,y)
 
We only "need" to cut for `regionRadiusSq` when computing the Wiener-filter, but not when performing smearing.

The problem with vectorization is that we have a varying box size in the innermost nested loop, which is not easy (if even possible) to vectorise.

I think we should focus on nailing the code, and sticking as close to the original as possible before optimizing it.

**Comment by alxndrwllmsn** [comment]
I'll create a new issue to do with the correctness of the output as it doesn't have anything to do with the optimisation. That way we can work on that in parallel.

In terms of optimisation, vectorisation improves the processing time by a factor of 8 to 16 (through testing in a notebook), so it is worthwhile doing. The varying box size is a solvable problem in that we can set the actual box size to the max that it will be and use masking to limit those that are smaller.

Alternatively, I can write the bulk of the processing in C++ to reduce the overhead induced by python and would be closer to the original code. This is a little more involved to do, so if we can get the vectorisation to work, I would prefer that.

**Comment by alxndrwllmsn** [comment]
I'm closing the issue. Since we've merged the branch.
